File: client/Ciphelet/app/src/main/java/com/github/worldoneo/ciphelet/connector/RegistrationService.java
package com.github.worldoneo.ciphelet.connector;

import androidx.core.util.Consumer;

import com.github.worldoneo.ciphelet.connector.action.GenericAction;
import com.github.worldoneo.ciphelet.connector.action.RegisterAction;
import com.github.worldoneo.ciphelet.connector.api.ChallengeHandler;
import com.github.worldoneo.ciphelet.connector.encryption.EncryptionUtility;
import com.iwebpp.crypto.TweetNaclFast;

public class RegistrationService {
    private final Connector connector;
    private Consumer<CipheletAPI> onRegister;

    public RegistrationService(Connector connector) {
        this.connector = connector;
    }

    public void onRegister(Consumer<CipheletAPI> register) {
        this.onRegister = register;
    }

    public void register(final String password) {
        connector.Connect();
        System.out.println("Register0");
        final TweetNaclFast.Box.KeyPair keyPair = EncryptionUtility.GenerateKeypair();
        GenericAction genericAction = new GenericAction("register");
        RegisterAction rAction = new RegisterAction();
        rAction.password = password;
        byte[] pKey = keyPair.getPublicKey();
        rAction.key = EncryptionUtility.encodeKey(pKey);
        genericAction.registerAction = rAction;
        final byte[] privateKey = keyPair.getSecretKey();
        connector.once(GenericAction.Action.REGISTER.response, new Consumer<GenericAction>() {
            @Override
            public void accept(GenericAction genericAction) {
                onRegister.accept(new CipheletAPI(genericAction.registerAction.humanid, connector, privateKey));
            }
        });

        connector.once(GenericAction.Action.CHALLENGE.response, new ChallengeHandler(privateKey, connector));
        connector.sendAction(genericAction);
    }
}

File: client/Ciphelet/app/src/main/java/com/github/worldoneo/ciphelet/connector/RegistrationService.kt
package com.github.worldoneo.ciphelet.connector

import androidx.core.util.Consumer
import com.github.worldoneo.ciphelet.connector.action.GenericAction
import com.github.worldoneo.ciphelet.connector.action.RegisterAction
import com.github.worldoneo.ciphelet.connector.api.ChallengeHandler
import com.github.worldoneo.ciphelet.connector.encryption.EncryptionUtility

class RegistrationService(private val connector: Connector, private val onRegister: Consumer<CipheletAPI>) {
    fun register(password: String) {
        connector.Connect()
        println("Register0")
        val keyPair = EncryptionUtility.GenerateKeypair()
        val genericAction = GenericAction("register")
        val rAction = RegisterAction(password)
        val pKey = keyPair.publicKey
        rAction.key = EncryptionUtility.encodeKey(pKey)
        genericAction.registerAction = rAction
        val privateKey = keyPair.secretKey
        connector.once(GenericAction.Action.REGISTER.response, Consumer { onRegister.accept(CipheletAPI(genericAction.registerAction!!.userid, connector, privateKey)) })
        connector.once(GenericAction.Action.CHALLENGE.response, ChallengeHandler(privateKey, connector))
        connector.sendAction(genericAction)
    }

}